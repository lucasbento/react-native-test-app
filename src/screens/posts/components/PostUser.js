import React from 'react';
import PropTypes from 'prop-types';
import {Text, StyleSheet} from 'react-native';

const PostUser = ({name, username, company}) => (
  <Text style={styles.user}>
    By {name} (<Text style={styles.username}>@{username}</Text>) from{' '}
    {company.name}
  </Text>
);

PostUser.propTypes = {
  name: PropTypes.string.isRequired,
  username: PropTypes.string.isRequired,
  company: PropTypes.shape({
    name: PropTypes.string,
  }).isRequired,
};

const styles = StyleSheet.create({
  user: {
    marginTop: 7,
    color: '#7E7E7E',
  },
  username: {
    color: '#00B2FF',
  },
});

export default PostUser;
