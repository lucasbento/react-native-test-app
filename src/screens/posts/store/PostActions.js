const PostActionTypes = {
  FETCH_POSTS_REQUEST: 'FETCH_POSTS_REQUEST',
  FETCH_POSTS_SUCCESS: 'FETCH_POSTS_SUCCESS',
  FETCH_POSTS_FAIL: 'FETCH_POSTS_FAIL',
  FETCH_POST_REQUEST: 'FETCH_POST_REQUEST',
  FETCH_POST_SUCCESS: 'FETCH_POST_SUCCESS',
  FETCH_POST_FAIL: 'FETCH_POST_FAIL',
  // This is used when the component unmounts so the next time
  // the user opens the `PostDetail` screen it doesn't blink on first render
  RESET_LAST_FETCHED_POST: 'RESET_LAST_FETCHED_POST',
};

const PostActions = {
  fetchPostsRequest: () => ({
    type: PostActionTypes.FETCH_POSTS_REQUEST,
  }),
  fetchPostsSuccess: ({posts}) => ({
    type: PostActionTypes.FETCH_POSTS_SUCCESS,
    payload: {
      posts,
    },
  }),
  fetchPostsFail: () => ({
    type: PostActionTypes.FETCH_POSTS_FAIL,
  }),
  fetchPostRequest: ({id, userId}) => ({
    type: PostActionTypes.FETCH_POST_REQUEST,
    payload: {
      id,
      userId,
    },
  }),
  fetchPostSuccess: ({post}) => ({
    type: PostActionTypes.FETCH_POST_SUCCESS,
    payload: {
      post,
    },
  }),
  fetchPostFail: () => ({
    type: PostActionTypes.FETCH_POST_FAIL,
  }),
  resetLastFetchedPost: () => ({
    type: PostActionTypes.RESET_LAST_FETCHED_POST,
  }),
};

export {PostActionTypes};
export default PostActions;
