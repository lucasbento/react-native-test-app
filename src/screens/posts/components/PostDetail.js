import React, {useEffect} from 'react';
import PropTypes from 'prop-types';
import {ScrollView, Text, StyleSheet} from 'react-native';
import {connect} from 'react-redux';
import PostActions from '../store/PostActions';
import Loading from '../../../components/Loading';
import PostUser from './PostUser';

const PostDetail = ({navigation, actions, isFetchPostLoading, post}) => {
  useEffect(() => {
    const id = navigation.getParam('id');
    const userId = navigation.getParam('userId');

    actions.fetchPostRequest({id, userId});
  }, [actions, navigation]);

  useEffect(
    () => () => {
      console.log(actions);
      // This is here for when the user opens this screen again it doesn't
      // blink on first render
      actions.resetLastFetchedPost();
    },
    [actions],
  );

  if (isFetchPostLoading) {
    return <Loading />;
  }

  const {title, body, user} = post;

  return (
    <ScrollView style={styles.container}>
      <Text style={styles.title}>{title}</Text>

      <PostUser
        name={user.name}
        username={user.username}
        company={user.company}
      />

      <Text style={styles.body}>{body}</Text>
    </ScrollView>
  );
};

PostDetail.navigationOptions = ({navigation}) => ({
  title: `Post ${navigation.getParam('id')}`,
});

PostDetail.propTypes = {
  navigation: PropTypes.shape({
    getParam: PropTypes.func,
  }).isRequired,
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    marginHorizontal: 10,
    marginTop: 10,
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
  },
  body: {
    marginTop: 20,
  },
});

const mapStateToProps = ({post}) => ({
  isFetchPostLoading: post.isFetchPostLoading,
  post: post.post,
});

const mapDispatchToProps = dispatch => ({
  actions: {
    fetchPostRequest: ({id, userId}) =>
      dispatch(PostActions.fetchPostRequest({id, userId})),
    resetLastFetchedPost: () => dispatch(PostActions.resetLastFetchedPost()),
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(PostDetail);
