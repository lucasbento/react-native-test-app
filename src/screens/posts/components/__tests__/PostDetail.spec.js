import React from 'react';
import {
  render,
  waitForElement,
  flushMicrotasksQueue,
} from '../../../../configuration/renderer';
import PostActions from '../../store/PostActions';
import PostDetail from '../PostDetail';
import PostUser from '../PostUser';
import post from '../../../../../__fixtures__/post';

describe('<PostDetail />', () => {
  const props = {
    navigation: {
      getParam: jest.fn(param => {
        const paramsToReturn = {
          id: post.id,
          userId: post.userId,
        };

        return paramsToReturn[param];
      }),
    },
  };

  it('renders correctly', async () => {
    const {toJSON} = render(<PostDetail {...props} />);

    await flushMicrotasksQueue();

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders correctly when loading', async () => {
    const {toJSON} = render(<PostDetail {...props} />);

    expect(toJSON()).toMatchSnapshot();
  });

  it('renders post title', async () => {
    const {getByText} = render(<PostDetail {...props} />);

    const postTitle = await waitForElement(() => getByText(post.title));

    expect(postTitle).toBeDefined();
  });

  it('renders post body', async () => {
    const {getByText} = render(<PostDetail {...props} />);

    const postBody = await waitForElement(() => getByText(post.body));

    expect(postBody).toBeDefined();
  });

  it('renders post user', async () => {
    const {getByType} = render(<PostDetail {...props} />);

    const postUser = await waitForElement(() => getByType(PostUser));

    expect(postUser).toBeDefined();
  });

  it('resets the last fetched post when unmounting', async () => {
    const resetLastFetchedPost = jest.spyOn(
      PostActions,
      'resetLastFetchedPost',
    );
    const {unmount} = render(<PostDetail {...props} />);

    unmount();

    expect(resetLastFetchedPost).toHaveBeenCalled();
    expect(resetLastFetchedPost).toHaveBeenCalledTimes(1);
  });
});
