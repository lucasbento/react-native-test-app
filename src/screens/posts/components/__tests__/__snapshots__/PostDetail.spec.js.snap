// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostDetail /> renders correctly 1`] = `
<RCTScrollView
  style={
    Object {
      "flex": 1,
      "marginHorizontal": 10,
      "marginTop": 10,
    }
  }
>
  <View>
    <Text
      style={
        Object {
          "fontSize": 20,
          "fontWeight": "bold",
        }
      }
    >
      sunt aut facere repellat provident occaecati excepturi optio reprehenderit
    </Text>
    <Text
      style={
        Object {
          "color": "#7E7E7E",
          "marginTop": 7,
        }
      }
    >
      By 
      Leanne Graham
       (
      <Text
        style={
          Object {
            "color": "#00B2FF",
          }
        }
      >
        @
        Bret
      </Text>
      ) from
       
      Romaguera-Crona
    </Text>
    <Text
      style={
        Object {
          "marginTop": 20,
        }
      }
    >
      quia et suscipit
suscipit recusandae consequuntur expedita et cum
reprehenderit molestiae ut ut quas totam
nostrum rerum est autem sunt rem eveniet architecto
    </Text>
  </View>
</RCTScrollView>
`;

exports[`<PostDetail /> renders correctly when loading 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <ActivityIndicator
    animating={true}
    color="#999999"
    hidesWhenStopped={true}
    size="large"
  />
</View>
`;
